[Bug 800880] Add Fallback Local in wiki

safwanrahman opened the pull request:
* The Local is just for test. It will be changed upon Local mapping

r? for the rest other?

mythmon commented on kitsune/wiki/views.py:
```diff
@@ -125,6 +125,32 @@ def document(request, document_slug, template=None, document=None):
+    # If any local has fallback locals and the docmunet is missing localization in that local
+    # So show the fallback local's document. There can be highest two fallback local for any local
+    # Fallback locals can be add in kitsune/wiki/config.py
+    if fallback_reason == 'no_translation' and (request.LANGUAGE_CODE in FALLBACK_LOCALS.keys()):
```
This can be more simply written as `if fallback_reason == 'no_translation' and request.LANGUAGE_CODE in FALLBACK_LOCALS:`. The `in` operator when applied with a string and a dictionary returns true if any of the keys in that dictionary equal the string.

mythmon commented on kitsune/wiki/views.py:
```diff
@@ -125,6 +125,32 @@ def document(request, document_slug, template=None, document=None):
+    # Fallback locals can be add in kitsune/wiki/config.py
+    if fallback_reason == 'no_translation' and (request.LANGUAGE_CODE in FALLBACK_LOCALS.keys()):
+        # Get fallback locals for the requested local
+        fallback_local_code = FALLBACK_LOCALS.get(request.LANGUAGE_CODE)
```
This should probably be written as `FALLBACK_LOCALS[request.LANGUAGE_CODE]`. `.get()` is mainly used in case the key might not exist in the dictionary. You already know that the language code is in `FALLBACL_LOCALS`, thanks to the `if` statement above.

mythmon commented on kitsune/wiki/views.py:
```diff
@@ -125,6 +125,32 @@ def document(request, document_slug, template=None, document=None):
+            # Lets show the English version
+            else:
+                doc = get_object_or_404(Document, locale=settings.WIKI_DEFAULT_LANGUAGE,
+                                        slug=document_slug)
```
I would write this a loop so that there can be any number of fallback languages. Something like

```python
translated_locales = doc.translation.all().values_list('locale', flat=True)
for fallback_locale in fallback_locale_codes:
    if fallback_locale in translated_locales:
        doc = doc.translated_to(fallback_locale)
        break
else:
    doc = get_object_or_404(Document, locale=settings.WIKI_DEFAULT_LANGAUGE, slug=document_slug)
```

mythmon replied:
Note: the `for / else` construct works by running through the for loop, and then if it gets to the end without breaking, running the `else` clause. If the `break` is hit, the else clause is skipped.

safwanrahman commented:
@mythmon Thoughts about the code?
Only showing message in template is awaiting.
r? for this code?

mythmon commented on kitsune/wiki/views.py:
```diff
@@ -125,6 +125,30 @@ def document(request, document_slug, template=None, document=None):
+                translation = doc.translated_to(fallback_local)
+                doc = Document.objects.get(locale=fallback_local,
+                                           slug=translation.slug)
+                fallback_reason = 'fallback_local'   # For showing note to users
```
You probably want 'fallback_locale'.

mythmon commented on kitsune/wiki/views.py:
```diff
@@ -125,6 +125,30 @@ def document(request, document_slug, template=None, document=None):
+                fallback_reason = 'fallback_local'   # For showing note to users
+                break
+            else:
+                doc = get_object_or_404(Document, locale=settings.WIKI_DEFAULT_LANGUAGE,
```
This `else` needs to be de-dented by one left so that it is associated with the `for` on line 141, not the `if` on line 142. The way you have it here, only the first fallback locale will be considered.

mythmon commented:
The code looks fine, though you should likely test this. The logic is pretty complex. Of course there also needs to be the message to the user.

That being said, I think the alternative proposed in the bug is a better idea. Putting this in the control of users to set their own fallback languages sounds like a better idea than trying to set the right fallback language in the code for every possible language.

safwanrahman commented:
@mythmon I have updated the code as mentioned in the bug. Sorry I squash it into ths last commit. It was a mistake! 
Can you please check the code? If it pass, I will start writing test.

mythmon commented on kitsune/wiki/templates/wiki/includes/document_macros.html:
```diff
@@ -71,6 +71,15 @@ <h1 class="title">{{ document.title }}</h1>
+          You can also see the <a href="{{ eng_link }}">English</a> version of this page.
+        {% endtrans %}
+      </div>
+        {{ document.html|safe }}
```
nit: this line is indented too much.

safwanrahman commented:
@mythmon, I have added test based on the patch. And corrected some Errors which was raised while testing this thing. I hope this should pass :pray: 
r?

safwanrahman commented:
@mythmon I have added one more test in templates. Also added comments in past test to make understand. I think its helpful. r?

mythmon commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,144 @@ def test_it_works(self):
+
+
+class FallbackSystem(TestCase):
+    "Check Article Wise Fallback Locale is working correclty"
```
Doc strings should use triple quotes. Plus the wording here is a bit awkward. Something like this would be better:

```python
"""Check per-article fallback locales are working correctly."""
```

mythmon commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,144 @@ def test_it_works(self):
+    def test_custom_locale_mapping_for_header_locale(self):
+        """Test while the document is not translated into any of the locale of ACCEPT_LANGUAGE
+        Header. But we have fallback locale mentioned in Custom wiki fallback locale for one
+        of the locale of ACCEPT_LANGUAGE Header"""
```
Doc strings should have a short summary in one line. If there needs to be more information, it should go after a blank line for example:

```python
"""
Test that configured locale fallbacks are used.

With a document not translated into any of the locales available in
ACCEPT_LANGUAGE, test that locales listed as fallbacks in `wiki/config.py` are
used.
"""
```

mythmon commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,144 @@ def test_it_works(self):
+        # custom wiki locale mapping for pt-PT. the fallback locale is pt-BR. and the document is
+        # translated into pt-BR, so the user should see pt-BR vesion of the document.
+        en_content = 'This artcile is in English'
+        trans_content = 'This artcile is translated into pt-BR'
```
You have a mispelling here (and in many other places). "artcile" should be "article". Please fix this everywhere it occurs instead of repeating the spelling mistake to make the tests pass.

mythmon commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,144 @@ def test_it_works(self):
+
+        # While the Requested locale not in Custom Wiki fallback locale
+        # A Client with ca, bn-IN, ja language choices in ACCEPT_LANGUAGE Header
+        header = 'ar,bn-IN;q=0.7,ja;q=0.3,'
```
This is an independent test case, so it should be in a separate test function.

safwanrahman replied:
Do you have any suggestion about the test function name?

mythmon commented on kitsune/wiki/views.py:
```diff
@@ -1445,3 +1473,111 @@ def what_links_here(request, document_slug):
+    return fallback_locale
+
+
+# Import from django.utils.translation.trans_real and changed some
```
"and changed some"... what? You should list the specific things you changed.

mythmon commented on kitsune/wiki/views.py:
```diff
@@ -1445,3 +1473,111 @@ def what_links_here(request, document_slug):
+    if browser_language_code is not None:
+        header_locales = parse_accept_lang_header(browser_language_code)
+    else:
+        header_locales = ''
```
This is bad. In one branch of the if, `header_locales` is a list (of tuples). In the other branch it is a string. It is bad that the type varies based on the branch. This line should be `header_locales = []`.

mythmon commented on kitsune/wiki/views.py:
```diff
@@ -1445,3 +1473,111 @@ def what_links_here(request, document_slug):
+        # This locale mapping is mentioned in kitsune/settings.py
+        for locale, __ in header_locales:
+            if (locale in NON_SUPPORTED_LOCALES and
+                    NON_SUPPORTED_LOCALES.get(locale) in translated_locales):
```
You can just say either `if NON_SUPPORTED_LOCALES.get(locale) in translated_locales`. Checking if something is in the dict and then using `.get()` is redundant. Either that or replace `.get(locales)` with `[locales]`.

mythmon commented on kitsune/wiki/views.py:
```diff
@@ -1445,3 +1473,111 @@ def what_links_here(request, document_slug):
+        # If any locale of Accpeted_Language Header is not in translated_locale,
+        # Check if we have any fallback local for the locales of Accpeted_Languge Header
+        # This locale mapping is mentioned in kitsune/settings.py
+        for locale, __ in header_locales:
```
Using two underscores is uncommon. It is more common to use one. `__` vs `_`. Using unexpected things makes people worry about the wrong things.

mythmon replied:
Also, since you have a modified version of `parse_accept_lang_header` anyways, you should make it only return the part you need (the locale) instead of returning a tuple.

mythmon commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,144 @@ def test_it_works(self):
+        # there is fr locale and the document is translated into fr locale, the user will see
+        # fr version of the document.
+        en_content = 'This artcile is in English'
+        trans_content = 'This artcile is translated into fr'
```
"artcile" should be "article".

safwanrahman commented:
@mythmon I have corrected the Docstring and comments. And change in views as per mentioned.
r?

vesper-mozilla commented on kitsune/wiki/config.py:
```diff
@@ -28,6 +28,20 @@
+    'gu-IN': ['hi-IN'],  # Bug 800880
+    'wo': ['fr'],  # Bug 800880
+    'fy-NL': ['nl'],  # Bug 800880
+    'wo': ['fr'],  # Bug 800880
```
Duplicate of line 37, please remove.

vesper-mozilla commented on kitsune/wiki/config.py:
```diff
@@ -28,6 +28,20 @@
+    'wo': ['fr'],  # Bug 800880
+    'pt-PT': ['pt-BR'],  # Bug 800880
+    'ta-LK': ['ta'],  # Bug 800880
+    'zh-CN': ['zh-TW'],  # Bug 800880
```
Please change to:

'zh-HK': ['zh-TW'],  # Bug 800880

as per Peter's comment (https://bugzilla.mozilla.org/show_bug.cgi?id=800880#c27)

vesper-mozilla commented on kitsune/wiki/config.py:
```diff
@@ -28,6 +28,20 @@
+    'pt-PT': ['pt-BR'],  # Bug 800880
+    'ta-LK': ['ta'],  # Bug 800880
+    'zh-CN': ['zh-TW'],  # Bug 800880
+    'zh-TW': ['zh-CN'],  # Bug 800880
```
Please remove as per Peter's comment (https://bugzilla.mozilla.org/show_bug.cgi?id=800880#c27)

vesper-mozilla commented on kitsune/wiki/templates/wiki/includes/document_macros.html:
```diff
@@ -71,6 +71,15 @@ <h1 class="title">{{ document.title }}</h1>
+        {% trans locale=full_locale_name[request.LANGUAGE_CODE], fallback_locale=full_locale_name[document.locale], help_link=url('wiki.document', 'localize-firefox-help'), eng_link=document.parent.get_absolute_url() %}
+          This page does not exist in {{ locale }}. You have been redirected to the {{ fallback_locale }} version instead.
+          If you would like to localize it into {{ locale }}, <a href="{{ help_link }}">click here</a>.
+          You can also see the <a href="{{ eng_link }}">English</a> version of this page.
```
Excuse my ignorance - is this text (starting with "This page does not exist...") localizable in Verbatim in its current format here? If not, can it be?

safwanrahman replied:
Yap. This text will surely localizable in Verbatim. In verbatim the localizer will see as following

```
This page does not exist in {{ locale }}. You have been redirected to the {{ fallback_locale }} version instead. If you would like to localize it into {{ locale }}, <a href="{{ help_link }}">click here</a>. You can also see the <a href="{{ eng_link }}">English</a> version of this page.
```

vesper-mozilla commented on kitsune/wiki/tests/test_templates.py:
```diff
@@ -380,6 +380,35 @@ def test_links_follow(self):
+        """The document template falls back to fallback locale if there is
+        custom wiki fallback mapping for the locale and the locale have no translation
+        exists."""
+        # Create a English Document and a pt-BR Document
```
# Create an English document and a pt-BR document

vesper-mozilla commented on kitsune/wiki/tests/test_templates.py:
```diff
@@ -380,6 +380,35 @@ def test_links_follow(self):
+        trans_doc = document(parent=en_rev.document, locale='pt-BR', save=True)
+        trans_rev = revision(document=trans_doc, save=True,
+                             content='Translated text.', is_approved=True)
+        # Make the revision as current revision of the document
```
# Mark the created revision as the current one for the document

vesper-mozilla commented on kitsune/wiki/config.py:
```diff
@@ -28,6 +28,20 @@
     ]
 }
+FALLBACK_LOCALES = {
+    'bn-IN': ['bn-BD'],  # Bug 800880
```
Please also add this line below:
'bn-BD': ['bn-IN'],  # Bug 800880
as per discussion with Swarnava

vesper-mozilla commented on kitsune/wiki/tests/test_templates.py:
```diff
@@ -380,6 +380,35 @@ def test_links_follow(self):
+        trans_doc.current_revision = trans_rev
+        trans_doc.save()
+
+        # Now get the pt-PT version of the document. As pt-PT has a fallback locale pt-BR
```
# Get the pt-PT version of the document.
# Resolve to the pt-BR version
# because pt-PT has pt-BR set in FALLBACK_LOCALES in wiki/config.py

vesper-mozilla commented on kitsune/wiki/tests/test_templates.py:
```diff
@@ -380,6 +380,35 @@ def test_links_follow(self):
+        trans_doc.save()
+
+        # Now get the pt-PT version of the document. As pt-PT has a fallback locale pt-BR
+        # in wiki/config.py, it would should show pt-BR content to the user.
```
Please remove this

vesper-mozilla commented on kitsune/wiki/tests/test_templates.py:
```diff
@@ -380,6 +380,35 @@ def test_links_follow(self):
+        doc = pq(response.content)
+        eq_(trans_doc.title, doc('article h1.title').text())
+
+        # There will be a fallback message to user about this.
```
# Display fallback message to the user

vesper-mozilla commented on kitsune/wiki/tests/test_templates.py:
```diff
@@ -380,6 +380,35 @@ def test_links_follow(self):
+        # Removing this as it shows up in text(), and we don't want to depend
+        # on its localization.
+        doc('#doc-pending-fallback').remove()
+        # Check Included content is in the pt-BR
```
# Check that content is available in pt-BR

vesper-mozilla commented on kitsune/wiki/tests/test_templates.py:
```diff
@@ -380,6 +380,35 @@ def test_links_follow(self):
+
+        # There will be a fallback message to user about this.
+        eq_(1, len(doc('#doc-pending-fallback')))
+        # Removing this as it shows up in text(), and we don't want to depend
```
# Remove it to avoid dependence on localization of this element

(my coding ignorance doesn't help me figure out what the "it" part stands for, so this may require some additional context, sorry)

safwanrahman replied:
I think it should be as its now!

vesper-mozilla commented on kitsune/wiki/tests/test_templates.py:
```diff
@@ -380,6 +380,35 @@ def test_links_follow(self):
+        # There will be a fallback message to user about this.
+        eq_(1, len(doc('#doc-pending-fallback')))
+        # Removing this as it shows up in text(), and we don't want to depend
+        # on its localization.
```
Remove this line, unless it's needed for additional comments

safwanrahman replied:
Its needed

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        product(save=True)
+
+    def create_documents(self, locale):
+        """Create a English Document and a translated document of the locale"""
```
"""Create a document in English and a translated document for the locale"""

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+
+
+class FallbackSystem(TestCase):
+    """Check per-article fallback locales are working correctly."""
```
"""Check that fallback locales on article level are working correctly"""

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        en = settings.WIKI_DEFAULT_LANGUAGE
+        en_content = 'This article is in English'
+        trans_content = 'This article is translated into %slocale' % locale
+        # Create an English Article and a translation in the locale
```
# Create an English article and a translation for the locale

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        # Make the revision as current revision of the translated document
+        trans_doc.current_revision = trans_rev
+        trans_doc.save()
+        # return the english document and translated document
```
# Return both the English version and the localized version of the document

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        # Create a revision of the translated document
+        trans_rev = revision(document=trans_doc, content=trans_content,
+                             is_approved=True, save=True)
+        # Make the revision as current revision of the translated document
```
# Make the created revision the current one for the localized document

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        revision(document=en_doc, content=en_content, is_approved=True,
+                 is_ready_for_localization=True, save=True)
+        trans_doc = document(parent=en_doc, title='Translated to', locale=locale, save=True)
+        # Create a revision of the translated document
```
Create a new revision of the localized document

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        """Create documents and get provided locale document data"""
+        # Create a client with provided ACCEPT_LANGUAGE header information
+        client = Client(HTTP_ACCEPT_LANGUAGE=header)
+        # Create English and translated Document based on the create_doc_locale
```
Create an English and a localized version of the document, based on create_doc_locale

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        """
+        Test that the article is showing in the header locale
+
+        Test while the document is not translated to requested locale but translated to
```
Please change to "Test cases when the document is not localized to locale requested by the user, but is localized to a locale mentioned in the ACCEPT_LANGUAGE header" and delete redundant comment lines

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        Test while the document is not translated to requested locale but translated to
+        one of the locale mentioned in ACCEPT_LANGUAGE Header
+        """
+        # A Client with es, fr, ja language choices in ACCEPT_LANGUAGE Header
```
# Define a client with es, fr, ja as locale choices in the ACCEPT_LANGUAGE header

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        """
+        # A Client with es, fr, ja language choices in ACCEPT_LANGUAGE Header
+        header = 'es,fr;q=0.7,ja;q=0.3,'
+        # Create a Translated document in fr locale version.
```
# Create a document localized into fr

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        # A Client with es, fr, ja language choices in ACCEPT_LANGUAGE Header
+        header = 'es,fr;q=0.7,ja;q=0.3,'
+        # Create a Translated document in fr locale version.
+        # And try to get data of es locale vesion of document with above mentioned client
```
# Attempt to resolve to the es version of the document with the client defined before

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        doc_content = self.get_data_from_translated_document(header=header,
+                                                             create_doc_locale='fr',
+                                                             req_doc_locale='es')
+        # As the document is not translated into es locale version, but in the client header
```
# Show the fr version of the document based on existing client header, as it is not localized into es

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+                                                             create_doc_locale='fr',
+                                                             req_doc_locale='es')
+        # As the document is not translated into es locale version, but in the client header
+        # there is fr locale and the document is translated into fr locale, the user will see
```
Please delete this line and the one below

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+
+    def test_header_locales_not_in_translated_list(self):
+        """
+        Test the article is showing in the fallback locale of header locele
```
Entire section rewrite:
"Test that the article is showing in the header locale, with the following conditions:
* the article is not localized into the user-requested locale
* the article is not localized to any of the locales listed in the ACCEPT_LANGUAGE header
* there is a fallback locale defined for one of the locales listed in the the ACCEPT LANGUAGE header; it is set in kitsune/settings.py

Test that the fallback locale is used"

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        *** Test that the fallback locale is being used
+        """
+
+        # A Client with de, an, ja language choices in ACCEPT_LANGUAGE Header
```
# Define a client with de, an, ja as locale choices in the ACCEPT_LANGUAGE header

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+
+        # A Client with de, an, ja language choices in ACCEPT_LANGUAGE Header
+        header = 'de,an;q=0.7,ja;q=0.3,'
+        # Create a translated document in es locale version.
```
# Create a document localized into es

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        # A Client with de, an, ja language choices in ACCEPT_LANGUAGE Header
+        header = 'de,an;q=0.7,ja;q=0.3,'
+        # Create a translated document in es locale version.
+        # And try to get data of de locale vesion of document with above mentioned client
```
# Attempt to resolve to the de version of the document with the client defined before

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        doc_content = self.get_data_from_translated_document(header=header,
+                                                             create_doc_locale='es',
+                                                             req_doc_locale='de')
+        # As the document is not translated into de locale version, also there are no locale
```
Show the es version of the document based on the fallback locale for an set in kitsune/settings.py, as it is not localized into de and there is no available locale based on the client header

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        # As the document is not translated into de locale version, also there are no locale
+        # in client header which the document is translated. But we have a fallback locale
+        # mentioned in kitsune/settings.py for the an locale. it is es. As the document is
+        # translated into es version, the user should see es version of the document
```
Please delete this line and the two above as well.

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+
+    def test_custom_locale_mapping(self):
+        """
+        Test that configured locale fallbacks are used for requested locale
```
Entire section rewrite:

"Test that the article is showing according to defined locale fallback options (more than one), with the following conditions:

    the article is not localized into the user-requested locale
    the article is not localized into any of the locales listed in the ACCEPT_LANGUAGE header
    none of the ACCEPT_LANGUAGE header locales has a defined fallback locale
    the user-requested locale has more than one fallback locale defined and the requested article is localized into one of the defined fallback locales
    the fallback locale is set in kitsune/settings.py

Test that the article is shown in the defined fallback locale"

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        *** Test that the Article is showing in the configured fallback locale
+        """
+
+        # Create a translated document in bn-BD locale and try to get bn-IN version
```
# Create a document localized into bn-BD
# Attempt to resolve to the bn-IN version of the document without defining locale options for the client

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        doc_content = self.get_data_from_translated_document(header=None,
+                                                             create_doc_locale='bn-BD',
+                                                             req_doc_locale='bn-IN')
+        # As the Document is not translated into bn-IN and there is no locale of
```
# Show the bn-BD version of the document based on the fallback locale, as it is not localized into bn-IN and there is no available locale from the ACCEPT_LANGUAGE header

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        # As the Document is not translated into bn-IN and there is no locale of
+        # ACCEPT_LANUGAE header which the document to, But there is a custom wiki fallback
+        # mapping available for bn-IN. the fallback locale is bn-BD. So the user should
+        # see the document in bn-BD.
```
Please delete this line and the two above as well.

vesper-mozilla commented on kitsune/wiki/views.py:
```diff
@@ -1445,3 +1473,116 @@ def what_links_here(request, document_slug):
+            priority = 1.0
+        result.append((lang, priority))
+    result.sort(key=lambda k: k[1], reverse=True)
+    # Changed here to get the locale name only
```
# Get the locale name only

safwanrahman replied:
As mentioned above, need to mention about change!

vesper-mozilla commented on kitsune/wiki/views.py:
```diff
@@ -1445,3 +1473,116 @@ def what_links_here(request, document_slug):
+        ''', re.VERBOSE)
+    result = []
+    pieces = accept_language_re.split(lang_string)
+    print pieces   # Changed here. Removed lower()
```
I don't know if this comment is necessary.

safwanrahman replied:
This is necessary because I have changed this. According to change in Open source code, need to mention the change log. :smile:

vesper-mozilla commented on kitsune/wiki/views.py:
```diff
@@ -1445,3 +1473,116 @@ def what_links_here(request, document_slug):
+#  * As we need only the locale name, added code to get only locale name
+def parse_accept_lang_header(lang_string):
+    """
+    Parses the lang_string, which is the body of an HTTP Accept-Language
```
Entire section rewrite:
"Parse the lang_string from the body of the HTTP Accept-Language header
 Return a list of languages, ordered by their 'q' values
 Return an empty list if there are any format errors in lang_string"

safwanrahman replied:
This should not be changed because its here as its in the library. From where I have imported

vesper-mozilla commented on kitsune/wiki/views.py:
```diff
@@ -1445,3 +1473,116 @@ def what_links_here(request, document_slug):
+    return fallback_locale
+
+
+# Import from django.utils.translation.trans_real and changed as below
```
Entire section rewrite:
"# Imported from django.utils.translation.trans_real with the following changes:
 # Removed lower() to get locale name as required
 # Added code to get only the locale name"

vesper-mozilla commented on kitsune/wiki/views.py:
```diff
@@ -1445,3 +1473,116 @@ def what_links_here(request, document_slug):
+                            continue
+                        break
+                else:
+                    # If all fails, return None as fallback Locale
```
# If all fails, return 'None' as fallback locale

safwanrahman replied:
In Code, 'None' is different from None!

vesper-mozilla commented on kitsune/wiki/views.py:
```diff
@@ -1445,3 +1473,116 @@ def what_links_here(request, document_slug):
+            else:
+                for locale in header_locales:
+                    if locale in FALLBACK_LOCALES:
+                        # Check if the document is trnaslated into the fallback locale
```
# Check if the document is localized into the fallback locale

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        assert trans_content in doc_content
+
+    def test_custom_locale_mapping_for_header_locale(self):
+        """
```
Entire section rewrite:
"Test that defined locale fallbacks are used as header locales, with the following conditions:

    the article is not localized into the user-requested locale
    the article is not localized to any of the locales listed in the ACCEPT_LANGUAGE header
    the header locales do not have a defined fallback locale OR the article is not localized into any of those defined fallback locales
    the user-requested locale has a defined fallback locale, but the article is not localized into it
    one of the locales listed in the ACCEPT_LANGUAGE headers has a defined fallback locale and the article is localized into that fallback locale

Test that the defined fallback locale is used"

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        *** Test that the Article is showing in the configured fallback locale
+        """
+
+        # While the requested locale also in custom wiki fallback locale but none of the
```
Entire section rewrite:
# The user-requested locale has a defined custom wiki fallback locale, but the article is not localized into that fallback locale

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        # While the requested locale also in custom wiki fallback locale but none of the
+        # fallback locale for requested locale is translated
+
+        # A Client with ca, pt-PT, ja language choices in ACCEPT_LANGUAGE Header
```
# Define a client with ca, pt-PT, ja as locale choices in the ACCEPT_LANGUAGE header

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+
+        # A Client with ca, pt-PT, ja language choices in ACCEPT_LANGUAGE Header
+        header = 'ca,pt-PT;q=0.7,ja;q=0.3,'
+        # Create a translated document in pt-BR and try to get ca locale version of the
```
Entire section rewrite:
# Create a document localized into pt-BR
# Attempt to resolve to the ca version of the document with the client defined before

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        doc_content = self.get_data_from_translated_document(header=header,
+                                                             create_doc_locale='pt-BR',
+                                                             req_doc_locale='ca')
+        # As the ca version of the document is not translated moreover the we have
```
Entire section rewrite:
# Show the pt-BR version of the document based on existing custom wiki locale mapping for pt-PT, as it is not localized into ca

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        assert en_content not in doc_content
+        assert trans_content in doc_content
+
+        # While the Requested locale not in Custom Wiki fallback locale
```
Entire section rewrite:
# If the user-requested locale does not have a custom wiki fallback locale
# Define a client with ca, bn-IN, ja as locale choices in the ACCEPT_LANGUAGE header

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        # While the Requested locale not in Custom Wiki fallback locale
+        # A Client with ca, bn-IN, ja language choices in ACCEPT_LANGUAGE Header
+        header = 'ar,bn-IN;q=0.7,ja;q=0.3,'
+        # Create a translated document into bn-BD and try to get the ar locale version
```
Entire section rewrite:
# Create a document localized into bn-BD
# Attempt to resolve to the ar version of the document with the client defined before

vesper-mozilla commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -684,3 +685,170 @@ def test_it_works(self):
+        doc_content = self.get_data_from_translated_document(header=header,
+                                                             create_doc_locale='bn-BD',
+                                                             req_doc_locale='ar')
+        # As the ar version of the document is not translated, moreover the document is translated
```
Entire section rewrite:
# Show the article in bn-BD based on the custom wiki fallback locale mapping for bn-IN, as it is localized neither into any of the locales listed in the ACCEPT_LANGUAGE header, nor into ar

vesper-mozilla commented on kitsune/wiki/views.py:
```diff
@@ -125,6 +128,30 @@ def document(request, document_slug, template=None, document=None):
             # and OK to fall back to parent (parent is approved).
             fallback_reason = 'no_translation'
 
+    # If any document is not translated into requested locale, we will try to find out the
```
Entire section rewrite:
# Find and show the defined fallback locale rather than the English version of the document
# The fallback locale is defined based on the ACCEPT_LANGUAGE header, site-wide locale mapping and custom fallback locale
# The custom fallback locale is defined in the FALLBACK_LOCALES array in kitsune/wiki/config.py
# See bug 800880 for more details

vesper-mozilla commented on kitsune/wiki/views.py:
```diff
@@ -125,6 +128,30 @@ def document(request, document_slug, template=None, document=None):
+    if fallback_reason == 'no_translation':
+        fallback_locale = get_fallback_locale(doc, request)
+
+        # If there is any falllback locale, show in that locale.
```
# If a fallback locale is defined, show the document in that locale

vesper-mozilla commented on kitsune/wiki/views.py:
```diff
@@ -125,6 +128,30 @@ def document(request, document_slug, template=None, document=None):
+
+        # If there is any falllback locale, show in that locale.
+        if fallback_locale is not None:
+            # Get the fallback Locale and show doc in the locale
```
# Set the fallback locale and show the document in that locale

vesper-mozilla commented on kitsune/wiki/views.py:
```diff
@@ -125,6 +128,30 @@ def document(request, document_slug, template=None, document=None):
+            translation = doc.translated_to(fallback_locale)
+            doc = Document.objects.get(locale=fallback_locale,
+                                       slug=translation.slug)
+            # For showing message to users.
```
# Show the fallback locale explanation message to the user

vesper-mozilla commented on kitsune/wiki/views.py:
```diff
@@ -125,6 +128,30 @@ def document(request, document_slug, template=None, document=None):
+            fallback_reason = 'fallback_locale'
+            full_locale_name = {request.LANGUAGE_CODE: LOCALES[request.LANGUAGE_CODE].native,
+                                fallback_locale: LOCALES[fallback_locale].native}
+        # If no fallback locale is found, show in default English version
```
# If there is no defined fallback locale, show the document in English

vesper-mozilla commented on kitsune/wiki/views.py:
```diff
@@ -1445,3 +1473,116 @@ def what_links_here(request, document_slug):
+def get_fallback_locale(doc, request):
+    """Get best fallback local based on locale Mapping"""
+
+    # Get the Translated locales which have current revision
```
# Get localized article versions with current revisions

vesper-mozilla commented on kitsune/wiki/views.py:
```diff
@@ -1445,3 +1473,116 @@ def what_links_here(request, document_slug):
+    translated_locales = (doc.translations.all()
+                                          .values_list('locale', flat=True)
+                                          .exclude(current_revision=None))
+    # Get the Accepted_language header from user request
```
# Get the ACCEPT_LANGUAGE header from the user's request

vesper-mozilla commented on kitsune/wiki/views.py:
```diff
@@ -1445,3 +1473,116 @@ def what_links_here(request, document_slug):
+        header_locales = parse_accept_lang_header(browser_language_code)
+    else:
+        header_locales = []
+    # Get the Locale mapping from setting
```
# Get the locale mapping from the Kitsune settings

vesper-mozilla commented on kitsune/wiki/views.py:
```diff
@@ -1445,3 +1473,116 @@ def what_links_here(request, document_slug):
+    # Get the Locale mapping from setting
+    NON_SUPPORTED_LOCALES = settings.NON_SUPPORTED_LOCALES
+
+    # Check if the document is translated to any of the locale listed in Accepted_Language Header
```
# Check if the document is localized into any of the locales listed in the ACCEPT_LANGUAGE header

vesper-mozilla commented on kitsune/wiki/views.py:
```diff
@@ -1445,3 +1473,116 @@ def what_links_here(request, document_slug):
+            fallback_locale = locale
+            break
+    else:
+        # If any locale of Accpeted_Language Header is not in translated locale,
```
Entire section rewrite:
# If there are no versions matching the locales listed in the ACCEPT_LANGUAGE header, check for fallback locales defined for locales listed in the ACCEPT_LANGUAGE header
# If the document is available in a fallback locale, show that locale
# The fallback locales are defined in the FALLBACK_LOCALES array in kitsune/wiki/config.py

vesper-mozilla commented on kitsune/wiki/views.py:
```diff
@@ -1445,3 +1473,116 @@ def what_links_here(request, document_slug):
+                fallback_locale = NON_SUPPORTED_LOCALES[locale]
+                break
+        else:
+            # If above does not match, then check if we have custom Wiki Locale mapping
```
Entire section rewrite:
# Check for custom wiki locale mapping for the user-request locale if the above did not match
# The custom wiki locale mapping is defined in kitsune/wiki/config.py

vesper-mozilla commented on kitsune/wiki/views.py:
```diff
@@ -1445,3 +1473,116 @@ def what_links_here(request, document_slug):
+                        fallback_locale = locale
+                        break
+                else:
+                    # If above does not match, then we would check if any locale of
```
Entire section rewrite:
# If the above did not match, check for matches of any of the ACCEPT_LANGUAGE header locales in the custom wiki locale mapping

vesper-mozilla commented on kitsune/wiki/views.py:
```diff
@@ -1445,3 +1473,116 @@ def what_links_here(request, document_slug):
+                    # Accepted_Language Header is in the Custom Wiki Locale mappping
+                    for locale in header_locales:
+                        if locale in FALLBACK_LOCALES:
+                            # Check if the document is trnaslated into the fallback locale
```
# Check if the document is localized into the fallback locale

vesper-mozilla commented on kitsune/wiki/views.py:
```diff
@@ -1445,3 +1473,116 @@ def what_links_here(request, document_slug):
+                        # If all fails, return None as fallback Locale
+                        fallback_locale = None
+
+            # If Custom Wiki Locale Mapping is not available for the locale user requesed, then we
```
Entire section rewrite:
# If the custom wiki locale mapping does not include the user-requested locale, check for custom wiki locale mapping set as fallback locales for the locales listed the ACCEPT_LANGUAGE header

safwanrahman commented:
@mythmon I have corrected the doc-strings and the comments as per mentioned. Can you please review now? Also rebased upon master to fix the merge conflicts. :smile: :walking:

mythmon commented on kitsune/wiki/views.py:
```diff
@@ -127,6 +130,28 @@ def document(request, document_slug, template=None, document=None):
+            # Get the fallback Locale and show doc in the locale
+            translation = doc.translated_to(fallback_locale)
+            doc = Document.objects.get(locale=fallback_locale,
+                                       slug=translation.slug)
```
This has the same problem you're other PR did. You're fetching the document from the DB twice in a row.

mythmon commented on kitsune/wiki/views.py:
```diff
@@ -1453,3 +1479,116 @@ def what_links_here(request, document_slug):
+        for locale in header_locales:
+            if NON_SUPPORTED_LOCALES.get(locale) in translated_locales:
+                fallback_locale = NON_SUPPORTED_LOCALES[locale]
+                break
```
In this case, it is clearer to return early, like this

```python
for locale in header_locales:
    if NON_SUPPORTED_LOCALES.get(locale) in translated_locales:
        return NON_SUPPORTED_LOCALES[locale]

#  <note this is dedented, and there is no else clause>
# Check for custom wiki locale mapping for the user-request locale if
# the above did not match.
# The custom wiki locale mapping is defined in kitsune/wiki/config.py
```

mythmon replied:
You should apply this pattern of returning early to the rest of the function as well. It should greatly improve the clarity.

mythmon commented on kitsune/wiki/views.py:
```diff
@@ -1453,3 +1479,116 @@ def what_links_here(request, document_slug):
+    """Get best fallback local based on locale Mapping"""
+
+    # Get localized article versions with current revisions
+    translated_locales = (doc.translations.all()
```
`.all()`, when combine with `.exclude()` or `.filter()` does nothing.

mythmon commented on kitsune/wiki/views.py:
```diff
@@ -1453,3 +1479,116 @@ def what_links_here(request, document_slug):
+                                          .values_list('locale', flat=True)
+                                          .exclude(current_revision=None))
+    # Get the ACCEPT_LANGUAGE header from the user's request
+    browser_language_code = request.META.get('HTTP_ACCEPT_LANGUAGE', None)
```
The comment here only restates the code. It doesn't add anything.

`None` is the default for `.get()`. You can leave it off.

mythmon commented:
Your `get_fallback_locale` function is really complicated, and hard to read. I stared it for a bit, and rewrote most of it. I committed to a branch on my fork: https://github.com/mythmon/kitsune/commit/6289cc9147838538d4efd46e9e35a73fdf544f77. You should be able to pull from the branch if you'd like to incorporate the changes.

My version passes all the tests, which I think exercise the requirements well. It cheats a little bit, and deals with each locale in the `ACCEPT_LANGUAGES` header all at once. Yours checks all the accept-languages for translations, then non-supported, then fallbacks. Do you think that is needed? Would be better better to loop through the header 3 times?

Let me know if you see any problems with my version.

safwanrahman commented:
@mythmon Thanks thanks a lot for rewriting the ``fallback_locale`` function in simplified. I am glad to see the code. I have made some very little changes in the code and also fixed rest other issues as you mentioned in the comments. Also correcting the tests to use factory-boy.
Also rebased upon master to remove the merge conflict
r?

mythmon commented:
Travis doesn't like your changes. :boom:

safwanrahman commented:
@mythmon After all, Travis Passed! r?

mythmon commented:
This doesn't meet all the criteria in the bug, but we talked about how we are going to defer the requirements about documents being out-of-date. I think the code here covers all the other requirements, and it isn't too messy.

Lets merge this, and work on the rest another time.
